extra_data/write_cxi.py

```diff
@@ -92,46 +134,78 @@ def _map_chunk(self, chunk, chunk_data, target, tgt_ax1, have_data=None):
                 f"{target_tids[0]} != {chunk_tids[0]}"
 
             # How much of this chunk can be mapped in one go?
-            mismatches = (chunk_tids != target_tids).nonzero()[0]
+            mismatches = (chunk_tids_target != target_tids).nonzero()[0]
             if mismatches.size > 0:
-                n_match = mismatches[0]
+                n_match = int(mismatches[0] / self.ncells)
             else:
                 n_match = len(chunk_tids)
 
             # Select the matching data and add it to the target
             chunk_match_end = chunk_match_start + n_match
+            tgt_end = tgt_start + (n_match*self.ncells)
             matched = chunk_data[chunk_match_start:chunk_match_end]
-            target[tgt_start : tgt_start+n_match, tgt_ax1] = matched
+            if self.ncells == 1:
+                # In some cases, there's an extra dimension of length 1.
+                # E.g. Jungfrau data with 1 memory cell per train or
+                # AGIPD/LPD raw data.
+                # Also, 'VirtualSource' object has no attribute 'ndim'.
+                if (len(matched.shape) > 1 and matched.shape[1] == 1):
+                    target[tgt_start:tgt_end, tgt_ax1] = matched[:, 0]
+                else:
+                    target[tgt_start:tgt_end, tgt_ax1] = matched
+            else:
+                for tid in range(n_match):
+                    start = tgt_start + tid*self.ncells
+                    end = tgt_start + (tid + 1)*self.ncells
+                    target[start:end, tgt_ax1] = matched[tid, :]
 
             # Fill in the map of what data we have
             if have_data is not None:
-                have_data[tgt_start : tgt_start+n_match, tgt_ax1] = True
+                have_data[tgt_start:tgt_end, tgt_ax1] = True
 
             # Prepare remaining data in the chunk for the next match
             chunk_match_start = chunk_match_end
             chunk_tids = chunk_tids[n_match:]
 
     def collect_data(self):
+        """
+        Prepare virtual layouts and map them to the virtual sources in
+        the data chunks.
+
+        Returns
+        -------
+
+        layouts: dict
+          A dictionary mapping virtual datasets names (e.g. ``data``)
+          to h5py virtual layouts.
+        """
         src = next(iter(self.detdata.source_to_modno))
         h5file = self.data._source_index[src][0].file
-        image_grp = h5file['INSTRUMENT'][src]['image']
+        image_grp = h5file['INSTRUMENT'][src][self.group_label]
 
         VLayout = h5py.VirtualLayout
 
         if 'gain' in image_grp:
-            log.info("Identified calibrated data")
+            log.info("Identified calibrated or Jungfrau raw data")
             shape = (self.nframes, self.nmodules) + self.detdata.module_shape
             log.info("Virtual data shape: %r", shape)
 
             layouts = {
-                'data': VLayout(shape, dtype=image_grp['data'].dtype),
+                self.image_label: VLayout(
```

Comment:
The name `data` in the output is actually part of the [CXI file format](https://raw.githubusercontent.com/cxidb/CXI/master/cxi_file_format.pdf), so I'd suggest keeping it the same regardless of the names in the input.

(I'm not sure whether we're actually producing valid CXI files - CrystFEL just uses whatever HDF5 dataset you tell it to - but as we're calling it CXI, we should try to follow the spec)

Reply:
I thought to preserve 'self.image_label' as a key for the layout since it is used later on as a key to access data from the HD5 file, currently it is in line 196:
```python3
for name, layout in layouts.items():
    key = '{}.{}'.format(self.group_label, name)
```
but still a virtual dataset is created with the name 'data', currently lines 287-289:
```python3
data = dgrp.create_virtual_dataset(
    'data', layouts[data_label], fillvalue=_fillvalues[data_label]
)
```
but please let me know if you would like also to name a key in layouts as 'data' - then I'll add an if statement to the line 196.

Reply:
Aha, OK. I was assuming that the keys in `layouts` matched the names the virtual datasets would have in the output file, rather than the names of the source datasets. There's no particular reason that has to be the case, though.